src/cluster.c

```diff
@@ -3651,6 +3652,19 @@ void clusterCron(void) {
         }
     }
 
+    /* Iterate nodes to remove free space in the send buffer
+    * to reclaim memory. */
+    di = dictGetSafeIterator(server.cluster->nodes);
+    while((de = dictNext(di)) != NULL) {
+        clusterNode *node = dictGetVal(de);
+        /* Resize the send buffer if it is wasting
+        * enough space. */
+        if (node->link != NULL && sdsavail(node->link->sndbuf) > SNDBUF_AVAILABLE_THRESHOLD) {
```

Comment:
I agree that because of the use of `sdscatlen` on `sndbuf` which internally can grow the buffer and create extra available space we might run into cases where we're growing and shrinking `sndbuf` back and forth. It turns into a race between `SNDBUF_AVAILABLE_THRESHOLD` and `sdsMakeRoomFor`'s internal implementation. We indeed encountered similar issues with the querybuf. To avoid this a simple assumption can be that if the available space is more than twice the size of the used space then we can shrink. Note that we should avoid using multiplication here because of (very theoretical) size overflows.

```suggestion
        /* If available space is more than twice the size of the buffer then free up unused space */
        if (node->link != NULL && sdsavail(node->link->sndbuf) / 2 > sdslen(node->link->sndbuf)) {
```

Reply:
+1